test/integration/filters/test_socket_interface.cc

```diff
@@ -0,0 +1,93 @@
+#include "test/integration/filters/test_socket_interface.h"
+
+#include <algorithm>
+
+#include "envoy/common/exception.h"
+#include "envoy/extensions/network/socket_interface/v3/default_socket_interface.pb.h"
+#include "envoy/network/socket.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/utility.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+TestIoSocketHandle::~TestIoSocketHandle() {
+  TestSocketInterface::GetMutableSingleton().clearSocket(this);
+}
+
+void TestIoSocketHandle::setWritevOverride(WritevOverrideProc override_proc) {
+  absl::WriterMutexLock lock(&mutex_);
+  writev_override_ = override_proc;
+}
+
+Api::IoCallUint64Result TestIoSocketHandle::writev(const Buffer::RawSlice* slices,
+                                                   uint64_t num_slice) {
```

Comment:
nit: A readv override would also be useful, but I suppose it can be added the first time it is needed.

Reply:
Yes, we can add it later. Not needed for now.